### PR TITLE
WitnessVersion type

### PR DESCRIPTION
**Pull request by dr-orlovsky**
Part of Taproot epic (should be listed in https://github.com/rust-bitcoin/rust-bitcoin/issues/503)

### CODE REVIEW

File: src/util/address.rs [hunk 1]
```diff
@@ -168,22 +337,19 @@ impl Payload {
     /// Get a [Payload] from an output script (scriptPubkey).
     pub fn from_script(script: &script::Script) -> Option<Payload> {
         Some(if script.is_p2pkh() {
-            Payload::PubkeyHash(PubkeyHash::from_slice(&script.as_bytes()[3..23]).unwrap())
+            Payload::PubkeyHash(PubkeyHash::from_slice(&script.as_bytes()[3..23]).expect("hash engine broken"))
```
Comment (sgeisler) [inline src/util/address.rs:340]: The failure has nothing to do with the hash engine but with supplying a byte slice of a wrong length.

File: src/util/address.rs [hunk 2]
```diff
@@ -148,6 +152,171 @@ impl FromStr for AddressType {
...
+}
+
+impl Display for WitnessVersion {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", *self as u8)
+    }
+}
+
+
+impl FromStr for WitnessVersion {
```
Comment (sgeisler) [inline src/util/address.rs:209]: Is this actually useful? Where do we directly (de)serialize a witness version? Isn't it always part of some other data structure with its own encoding?

File: src/util/address.rs [hunk 3]
```diff
@@ -168,22 +337,19 @@ impl Payload {
...
     pub fn from_script(script: &script::Script) -> Option<Payload> {
         Some(if script.is_p2pkh() {
-            Payload::PubkeyHash(PubkeyHash::from_slice(&script.as_bytes()[3..23]).unwrap())
+            Payload::PubkeyHash(PubkeyHash::from_slice(&script.as_bytes()[3..23]).expect("hash engine broken"))
         } else if script.is_p2sh() {
-            Payload::ScriptHash(ScriptHash::from_slice(&script.as_bytes()[2..22]).unwrap())
+            Payload::ScriptHash(ScriptHash::from_slice(&script.as_bytes()[2..22]).expect("hash engine broken"))
         } else if script.is_witness_program() {
-            // We can unwrap the u5 check and assume script length
+            // We can unwrap assume script length
```
Comment (sgeisler) [inline src/util/address.rs:344]: I think the `and` has to stay in the comment for it to make any sense.

**Review by sgeisler — COMMENTED** [review]
Looks good and improves the API quite a bit imo. I left a few nits regarding some `expects` and comments.

File: src/util/address.rs (hunk 2 above)
Comment (apoelstra) [reply src/util/address.rs:209]: I could imagine it being useful somewhere, and it's not much extra code.

Comment (sgeisler) [reply src/util/address.rs:209]: Yeah, it's not too bad now that I noticed even without it the `UnparsableWitnessVersion` error variant can't be made more expressive.

File: src/util/address.rs (hunk 1 above)
Comment (dr-orlovsky) [reply src/util/address.rs:340]: Removed `expect` at all by using infallible `from_inner`

**Comment by dr-orlovsky** [comment]
Fixed @sgeisler suggestions & force-pushed. Ready for re-review

**Comment by sgeisler** [comment]
Does this really still have to be its own PR? Afaik it's included in #563 (idk if based on an up to date version) and fits in there quite well. It's a breaking change anyway.

**Comment by sanket1729** [comment]
I have left some feedback on #563 for this commit.

**Comment by dr-orlovsky** [comment]
Updated with review suggestions from https://github.com/rust-bitcoin/rust-bitcoin/pull/563

**Comment by sgeisler** [comment]
Again, would anyone be opposed to closing this in favor of merging it with #563? It's just so related and reviewing #563 together with the #562 changes seemed easy enough. #563 is also close to merge ready imo, maybe a third review would be nice and we need to decide if we want to release 0.26.1 before that, but there's nothing fundamentally blocking just merging #563 all at once after that.

**Comment by dr-orlovsky** [comment]
Agree, closing this now

**Comment by dr-orlovsky** [comment]
Reopened as #617

**Pull request closed without merging**